```
warning CS8002: Referenced assembly 'WebDriver, Version=3.12.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Servers\Kestrel\test\Interop.FunctionalTests\Interop.FunctionalTests.csproj]
warning CS8002: Referenced assembly 'Pomelo.EntityFrameworkCore.MySql, Version=2.1.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Mvc\benchmarkapps\BasicApi\BasicApi.csproj]
warning CS8002: Referenced assembly 'Pomelo.EntityFrameworkCore.MySql, Version=2.1.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Mvc\benchmarkapps\BasicViews\BasicViews.csproj]
2019-01-04T02:12:47.8847130Z D:\a\1\s/src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj : warning NU1503: Skipping restore for project 'D:\a\1\s/src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj'. The project file may be invalid or missing targets required for restore. [C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets]
SiteExtensions\LoggingBranch\LoggingBranch.csproj :
 warning NU1503: Skipping restore for project 'D:\a\1\s/src\SiteExtensions\LoggingBranch\LoggingBranch.csproj'.
 The project file may be invalid or missing targets required for restore. [C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]

test\Identity.FunctionalTests\Microsoft.AspNetCore.Identity.FunctionalTests.csproj(42,5): 
 error : Unable to find precompiled view file bin\Release\netcoreapp3.0\Microsoft.AspNetCore.Identity.UI.Views.V3.dll
2019-01-04T02:12:47.8853580Z D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20181214.1\modules\vstest\module.targets(102,5):
 error : Test group UnitTests/netcoreapp3.0 failed with exit code '1'. [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20181214.1\KoreBuild.proj]
 
2019-01-04T02:12:47.8853651Z     13 Warning(s)
2019-01-04T02:12:47.8853693Z     2 Error(s)
```

--- a/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs
+++ b/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs
@@ -58,7 +58,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
             Output = _http1Output;
         }
 
-        public PipeReader Input { get; private set; }
+        public PipeReader Input { get; }
 
         public bool RequestTimedOut => _requestTimedOut;
 

--- a/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs
+++ b/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs
@@ -54,10 +54,11 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
                 _context.TimeoutControl,
                 this);
 
+            Input = _context.Transport.Input;
             Output = _http1Output;
         }
 
-        public PipeReader Input => _context.Transport.Input;
+        public PipeReader Input { get; private set; }
 
         public bool RequestTimedOut => _requestTimedOut;
 

--- a/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
+++ b/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
@@ -154,7 +154,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 
         private async Task ProcessReceives()
         {
-            // Resolve `input` PipeWriter via the IDuplexPipe interface prior to loop start
+            // Resolve `input` PipeWriter via the IDuplexPipe interface prior to loop start for performance.
             var input = Input;
             while (true)
             {
@@ -240,7 +240,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 
         private async Task ProcessSends()
         {
-            // Resolve `output` PipeReader via the IDuplexPipe interface prior to loop start
+            // Resolve `output` PipeReader via the IDuplexPipe interface prior to loop start for performance.
             var output = Output;
             while (true)
             {

--- a/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
+++ b/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
@@ -154,13 +154,15 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 
         private async Task ProcessReceives()
         {
+            // Resolve `input` PipeWriter via the IDuplexPipe interface prior to loop start
+            var input = Input;
             while (true)
             {
                 // Wait for data before allocating a buffer.
                 await _receiver.WaitForDataAsync();
 
                 // Ensure we have some reasonable amount of buffer space
-                var buffer = Input.GetMemory(MinAllocBufferSize);
+                var buffer = input.GetMemory(MinAllocBufferSize);
 
                 var bytesReceived = await _receiver.ReceiveAsync(buffer);
 
@@ -171,9 +173,9 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
                     break;
                 }
 
-                Input.Advance(bytesReceived);
+                input.Advance(bytesReceived);
 
-                var flushTask = Input.FlushAsync();
+                var flushTask = input.FlushAsync();
 
                 var paused = !flushTask.IsCompleted;
 
@@ -238,9 +240,11 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 
         private async Task ProcessSends()
         {
+            // Resolve `output` PipeReader via the IDuplexPipe interface prior to loop start
+            var output = Output;
             while (true)
             {
-                var result = await Output.ReadAsync();
+                var result = await output.ReadAsync();
 
                 if (result.IsCanceled)
                 {
@@ -256,7 +260,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
                     await _sender.SendAsync(buffer);
                 }
 
-                Output.AdvanceTo(end);
+                output.AdvanceTo(end);
 
                 if (isCompleted)
                 {

